spec/acceptance/middleware_spec.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require_relative "../spec_helper"
+
+describe "Middleware" do
+  it "is used in Rails by default" do
+    skip unless defined?(Rails)
+
+    app = Class.new(Rails::Application) do
+      config.eager_load = false
+      config.logger = Logger.new(nil) # avoid creating the log/ directory automatically
+      config.cache_store = :null_store # avoid creating tmp/ directory for cache
+    end
+
+    app.initialize!
+    assert app.middleware.include?(Rack::Attack)
+  end
+
+  def app
+    Rack::Builder.new do
+      use Rack::Attack do
+        blocklist_ip("1.2.3.4")
+      end
```

Comment:
can we actually test that multiple instances of the middleware work together?

Maybe I'm missing something but I tried

```ruby
  def app
    Rack::Builder.new do
      use Rack::Attack do
        blocklist_ip("1.2.3.4")
      end
      use Rack::Attack do
        blocklist_ip("4.3.2.1")
      end

      run lambda { |_env| [200, {}, ['Hello World']] }
    end.to_app
  end

  it "can be configured via a block" do
    get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
    assert_equal 403, last_response.status

    get "/", {}, "REMOTE_ADDR" => "4.3.2.1"
    assert_equal 403, last_response.status
  end
```

and it didn't work 🤔 thoughts?

Reply:
Multiple instances on the same rack app are not working because of this guard clause https://github.com/rack/rack-attack/blob/b6ce502b1491ed4de057d1dc20b8391f8d7b3d19/lib/rack/attack.rb#L105

Reply:
ahh ok I see.. that was introduced to [guarantee idempotency](https://github.com/rack/rack-attack/pull/457). But with this PR, that's no longer accurate.

What I mean is that in the past it was safe to just have `Rack::Attack` execute once because all configs were the same, but with this PR we could have different configs of `Rack::Attack` and only the first one to run will execute, the rest will silently be skipped, right?

At this point I'm starting to doubt if the benefits of the PR are greater than its drawbacks, especially around generating confusion to end users and inflicting pain on them 🤔

Pros I see:
- Nicer way to config using the block rather than the `Rack::Attack` class.

Cons I see:
- There are now 2 ways to config `Rack::Attack` which can increase confusion, especially since you can't use both at the same time.
- Idempotency guard no longer works as expected, now only the config of the first middleware to run, will take place.
- Potential conflicts between the automatically added middleware to Rails apps and the usage of the middleware with the new block method.

WDYT? Please let me know your thoughts, I'm open to discuss it and change my mind 🙌

Reply:
I think this is still a good direction.

Ideally we retain compatibility with any existing configuration mechanism and redirect it to a per-instance configuration.

I don't think the idempotency fix is correct. It seems like XY problem.

Reply:
@ioquatix Do you have suggestions how to proceed?

I wish the middleware was not automatically added to rails apps some time ago 😐.

Reply:
I don't have time to provide detailed feedback until after RubyKaigi, but I can take a look then. Is it part of Rails' default stack?

Reply:
> Is it part of Rails' default stack?

Yes, https://github.com/rack/rack-attack/blob/b6ce502b1491ed4de057d1dc20b8391f8d7b3d19/lib/rack/attack/railtie.rb#L11-L15

> but I can take a look then

👍

Reply:
Maybe we can introduce some configs so that you either get the old way to config + idempotency + the railtie, OR you get the new way to config rack attack with the block + ability to have multiple middlewares added + no railtie?

What I mean is that I’m ok with both approaches, but I think the key here is not mixing both of them at the same time 🤔

Reply:
It's fairly straight forward to configure using a Railtie, so we could have a default middleware which gets configured via Rails' own configuration system, but that doesn't mean we can't have non-global instances.